pkg/sentry/kernel/msgqueue/msgqueue.go

```diff
@@ -0,0 +1,227 @@
+
+	// senders holds a queue of blocked message senders. Senders are notified
+	// when enough space is available in the queue to insert their message.
+	senders waiter.Queue
+
+	// receivers holds a queue of blocked receivers. Receivers are notified
+	// when a new message is inserted into the queue and can be received.
+	receivers waiter.Queue
+
+	// messages is a list of sent messages.
+	messages msgList
+
+	// sendTime is the last time a msgsnd was perfomed.
+	sendTime ktime.Time
+
+	// receiveTime is the last time a msgrcv was performed.
+	receiveTime ktime.Time
+
+	// changeTime is the last time the queue was modified using msgctl.
+	changeTime ktime.Time
+
+	// byteCount is the current number of message bytes in the queue.
+	byteCount uint64
+
+	// messageCount is the current number of messages in the queue.
+	messageCount uint64
+
+	// maxBytes is the maximum allowed number of bytes in the queue, and is also
+	// used as a limit for the number of total possible messages.
+	maxBytes uint64
+
+	// sendPID is the PID of the process that performed the last msgsnd.
+	sendPID int32
+
+	// receivePID is the PID of the process that performed the last msgrcv.
+	receivePID int32
+}
+
+// Message represents a message exchanged through a Queue via msgsnd(2) and
+// msgrcv(2).
+//
+// +stateify savable
+type Message struct {
+	msgEntry
+
+	// mType is an integer representing the type of the sent message.
+	mType int64
+
+	// mText is an untyped block of memory.
+	mText []byte
+
+	// mSize is the size of mText.
+	mSize uint64
+}
+
+// FindOrCreate creates a new message queue or returns an existing one. See
+// msgget(2).
+func (r *Registry) FindOrCreate(ctx context.Context, key ipc.Key, mode linux.FileMode, private, create, exclusive bool) (*Queue, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if !private {
+		queue, err := r.reg.Find(ctx, key, mode, create, exclusive)
+		if err != nil {
+			return nil, err
+		}
+
+		if queue != nil {
+			return queue.(*Queue), nil
+		}
+	}
+
+	// Check system-wide limits.
+	if len(r.reg.Objects) >= maxQueues {
+		return nil, syserror.ENOSPC
+	}
+
+	return r.newQueue(ctx, key, fs.FileOwnerFromContext(ctx), fs.FilePermsFromMode(mode))
+}
+
+// newQueue creates a new queue using the given fields. An error is returned if
+// there're no more available identifiers.
+//
+// Precondition: r.mu must be held.
+func (r *Registry) newQueue(ctx context.Context, key ipc.Key, creator fs.FileOwner, perms fs.FilePermissions) (*Queue, error) {
+	id, err := r.reg.NewID()
+	if err != nil {
+		return nil, err
+	}
+	q := &Queue{
+		registry:    r,
+		obj:         ipc.NewObject(r.userNS, id, key, creator, creator, perms),
+		sendTime:    ktime.ZeroTime,
+		receiveTime: ktime.ZeroTime,
+		changeTime:  ktime.NowFromContext(ctx),
+		maxBytes:    maxQueueBytes,
+	}
+	r.reg.Objects[id] = q
+	r.reg.KeysToIDs[key] = id
+
+	return q, nil
+}
+
+// Object returns a pointer to queue's ipc.Object, and the mutex used to
+// protect it.
```

Comment:
We document interface method implementations as ([example](https://cs.opensource.google/gvisor/gvisor/+/master:pkg/sentry/kernel/shm/shm.go;l=435?q=shm.MappedName&ss=gvisor%2Fgvisor)):

"Object implements ipc.Mechanism.Object.", followed by any implementation-specific details.

Since there's nothing noteworthy about this implementation, there's no need to repeat what the `Object` method does, and just referencing the interface is sufficient.

Reply:
Done. Also updated for other `Mechanism` implementations.